spec.bs

```diff
@@ -467,48 +467,119 @@ A <dfn export for=fencedframetype>size</dfn> is a struct with non-negative integ
+A <dfn export for=fencedframetype>fenced frame reporting map</dfn> is a [=map=] whose [=map/keys=]
+are {{FenceReportingDestination}}s and whose [=map/values=] are either `"pending"`, or [=maps=]
+whose [=map/keys=] are [=strings=] and whose [=map/values=] are [=urls=].
+
+A <dfn export for=fencedframetype>fenced frame reporting metadata</dfn> is a struct with the following [=struct/items=]:
+<dl export dfn-for="fenced frame reporting metadata">
+  : <dfn>fenced frame reporting map</dfn>
+  :: a [=fencedframetype/fenced frame reporting map=]
+
+  : <dfn>direct seller is seller</dfn>
+  :: a boolean
+</dl>
 
 <div algorithm>
-  In order to <dfn export>report an event</dfn>, run these steps:
+  In order to <dfn export>finalize a reporting destination</dfn>, given a [=fencedframetype/fenced
+  frame reporting map=] |reporting map|, a {{FenceReportingDestination}} |destination|, and a
+  [=map=] |destination map| whose [=map/keys=] are [=strings=] and whose [=map/values=] are
+  [=urls=], run these steps:
 
-  1. <span class=XXX>TODO: fill this in</span>
+  1. [=Assert=] that |reporting map|[|destination|] is `"pending"`.
+
+  1. [=map/Set=] |reporting map|[|destination|] to |destination map|.
 </div>
 
+A <dfn export for=fencedframetype>fenced frame reporter</dfn> is a struct with the following [=struct/items=]:
+<dl export dfn-for="fenced frame reporter">
+  : <dfn>fenced frame reporting map reference</dfn>
+  :: a reference to a [=fencedframetype/fenced frame reporting map=]
+     <span class=XXX>TODO: Handle pointers/references in a more spec-y way</span>
+
+  : <dfn>direct seller is seller</dfn>
+  :: a boolean
+
+  : <dfn>automatic beacon data</dfn>
+  :: null, or a struct with the following [=struct/items=]:
+     <dl dfn-for="automatic beacon data">
+       : <dfn>eventData</dfn>
+       :: a [=string=]
+
+       : <dfn>destination</dfn>
+       :: a [=list=] of {{FenceReportingDestination}}s
+     </dl>
+</dl>
+
 <div algorithm>
-  In order to <dfn export>report a private aggregation event</dfn>, run these steps:
+  In order to <dfn>report an event</dfn> using a [=fencedframetype/fenced frame reporter=]
+  |reporter| with a {{FenceReportingDestination}} |destination|, [=string=] |eventType|, and
+  [=string=] |eventData|, run these steps:
+
+  1. If |destination| is `"direct-seller"`:
+
+     1. If |reporter|'s [=fenced frame reporter/direct seller is seller=] is true, set
+        |destination| to `"seller"`.
+
+     1. Otherwise, set |destination| to `"component-seller"`.
+
+  1. Let |reporting map| be |reporter|'s [=fenced frame reporter/fenced frame reporting map
+     reference=].
+
+  1. If |reporting map|[|destination|] does not [=map/exist=], return.
+
+  1. If |reporting map|[|destination|] is `"pending"`, return and asynchronously wait until it is
+     no longer pending.
+     <span class=XXX>TODO: Specify the waiting properly.
```

Comment:
I think it would be more ideal if we could handle this TODO in this PR. Can we give the map some sort of callback holding mechanism that can be invoked with "the rest of these steps" later on, and gets called at the appropriate time? (https://html.spec.whatwg.org/multipage/urls-and-fetching.html#lazy-load-resumption-steps and its references might be a good example to go off of).